core/utils/index_utils.cpp

```diff
@@ -28,148 +28,73 @@
 
-      return lhs.meta->name < rhs.meta->name;
-    }
+  ConsolidationCandidate::ConsolidationCandidate(
+    iterator_t start,
+    iterator_t end) noexcept
+  : segments(start, end) {
 
-    return lhs.size < rhs.size;
-  }
+    initialized = true;
 
-  operator const irs::SubReader*() const noexcept { return reader; }
-
-  const irs::SubReader* reader;
-  const irs::SegmentInfo* meta;
-  size_t size;  // approximate size of segment without removals
-  double_t fill_factor;
-};
-
-struct ConsolidationCandidate {
-  using iterator_t = std::vector<SegmentStats>::const_iterator;
-  using range_t = std::pair<iterator_t, iterator_t>;
-
-  explicit ConsolidationCandidate(iterator_t i) noexcept : segments(i, i) {}
-
-  iterator_t begin() const noexcept { return segments.first; }
-  iterator_t end() const noexcept { return segments.second; }
-
-  range_t segments;
-  size_t count{0};
-  size_t size{0};           // estimated size of the level
-  double_t score{DBL_MIN};  // how good this permutation is
-};
-
-/// @returns score of the consolidation bucket
-double_t consolidation_score(const ConsolidationCandidate& consolidation,
-                             const size_t segments_per_tier,
-                             const size_t floor_segment_bytes) noexcept {
-  // to detect how skewed the consolidation we do the following:
-  // 1. evaluate coefficient of variation, less is better
-  // 2. good candidates are in range [0;1]
-  // 3. favor condidates where number of segments is equal to
-  // 'segments_per_tier' approx
-  // 4. prefer smaller consolidations
-  // 5. prefer consolidations which clean removals
-
-  switch (consolidation.count) {
-    case 0:
-      // empty consolidation makes not sense
-      return DBL_MIN;
-    case 1: {
-      auto& meta = *consolidation.segments.first->meta;
-
-      if (meta.docs_count == meta.live_docs_count) {
-        // singletone without removals makes no sense
-        return DBL_MIN;
-      }
+    //  Calculate initial cost
+    auto itr = start;
+    do
+    {
+      auto itrMeta = itr->meta;
 
-      // FIXME honor number of deletes???
-      // signletone with removals makes sense if nothing better is found
-      return DBL_MIN + DBL_EPSILON;
-    }
-  }
+      mergeBytes += itrMeta->byte_size;
+      skew = static_cast<double>(itrMeta->byte_size) / mergeBytes;
+      delCount += (itrMeta->docs_count - itrMeta->live_docs_count);
+      mergeScore = skew + (1.0 / (1 + delCount));
+      cost = mergeBytes * mergeScore;
 
-  size_t size_before_consolidation = 0;
-  size_t size_after_consolidation = 0;
-  size_t size_after_consolidation_floored = 0;
-  for (auto& segment_stat : consolidation) {
-    size_before_consolidation += segment_stat.meta->byte_size;
-    size_after_consolidation += segment_stat.size;
-    size_after_consolidation_floored +=
-      std::max(segment_stat.size, floor_segment_bytes);
+    } while (itr++ != end);
```

Comment:
Probably inconsequential, but it would suffice to calculate `skew`, `mergeScore` and `cost` once after the loop for the last element.